src/main/terrier.cpp

```diff
@@ -1,53 +1,50 @@
-#include "loggers/transaction_logger.h"
-#include "loggers/type_logger.h"
-#include "storage/data_table.h"
-#include "storage/record_buffer.h"
-#include "storage/storage_defs.h"
-#include "transaction/transaction_context.h"
+#include <unordered_map>
+#include <utility>
+#include "main/db_main.h"
 
-int main() {
+/*
+ * Define gflags configurations.
+ * This will expand to a list of code like:
+ * DEFINE_int32(port, 15721, "Terrier port (default: 15721)");
+ */
+#define __SETTING_GFLAGS_DEFINE__      // NOLINT
+#include "settings/settings_common.h"  // NOLINT
+#include "settings/settings_defs.h"    // NOLINT
+#undef __SETTING_GFLAGS_DEFINE__       // NOLINT
+
+// Construct settings param map from settings_defs.h
+static void ConstructParamMap(                                                                // NOLINT
+    std::unordered_map<terrier::settings::Param, terrier::settings::ParamInfo> &param_map) {  // NOLINT
+  /*
+   * Populate gflag values to param map.
+   * This will expand to a list of code like:
+   * param_map_.emplace(
+   *     terrier::settings::Param::port,
+   *     terrier::settings::ParamInfo(port, terrier::type::TransientValueFactory::GetInteger(FLAGS_port),
+   *                                  "Terrier port (default: 15721)",
+   *                                  terrier::type::TransientValueFactory::GetInteger(15721), is_mutable));
+   */
+
+#define __SETTING_POPULATE__           // NOLINT
+#include "settings/settings_common.h"  // NOLINT
+#include "settings/settings_defs.h"    // NOLINT
+#undef __SETTING_POPULATE__            // NOLINT
+}
+
+int main(int argc, char *argv[]) {
   // initialize loggers
-  try {
-    init_main_logger();
-    // initialize namespace specific loggers
-    terrier::storage::init_index_logger();
-    terrier::storage::init_storage_logger();
-    terrier::transaction::init_transaction_logger();
-    terrier::parser::init_parser_logger();
-    terrier::network::init_network_logger();
-    // Flush all *registered* loggers using a worker thread.
-    // Registered loggers must be thread safe for this to work correctly
-    spdlog::flush_every(std::chrono::seconds(DEBUG_LOG_FLUSH_INTERVAL));
-  } catch (const spdlog::spdlog_ex &ex) {
-    std::cout << "debug log init failed " << ex.what() << std::endl;  // NOLINT
-    return 1;
-  }
-  // log init now complete
-  LOG_TRACE("Logger initialization complete");
+  // Parse Setting Values
+  ::google::SetUsageMessage("Usage Info: \n");
+  ::google::ParseCommandLineFlags(&argc, &argv, true);
+  std::unordered_map<terrier::settings::Param, terrier::settings::ParamInfo> param_map;
 
   // initialize stat registry
   auto main_stat_reg = std::make_shared<terrier::common::StatisticsRegistry>();
 
-  LOG_INFO("Initialization complete");
-
-  terrier::network::TerrierServer terrier_server;
-  terrier_server.SetupServer().ServerLoop();
-
-  // shutdown loggers
-  spdlog::shutdown();
-  main_stat_reg->Shutdown(false);
+  ConstructParamMap(param_map);
```

Comment:
This piece of code is weird. Is it possible to generate the param map during the database bootstrap phase? I.e. `db.Bootstrap()`, `db.GetParams()`, `db.Init()`, `db.Run()`.

Reply:
No, because we don't want to expose command line arguments to DBMain.